File: path_match.go
package farp

import "strings"

// MatchPath checks whether a URL path matches a glob pattern.
//
// Pattern syntax:
//   - Literal segments match exactly: "/api/v1/users" matches "/api/v1/users"
//   - "*" matches exactly one path segment: "/api/*/status" matches "/api/users/status"
//   - "**" matches zero or more path segments: "/api/**" matches "/api", "/api/v1", "/api/v1/users/123"
func MatchPath(pattern, path string) bool {
	patternParts := splitPath(pattern)
	pathParts := splitPath(path)

	return matchParts(patternParts, pathParts)
}

func splitPath(p string) []string {
	p = strings.Trim(p, "/")
	if p == "" {
		return nil
	}
	return strings.Split(p, "/")
}

func matchParts(pattern, path []string) bool {
	pi, pa := 0, 0

	for pi < len(pattern) && pa < len(path) {
		seg := pattern[pi]

		switch {
		case seg == "**":
			// If ** is the last pattern segment, it matches everything remaining
			if pi == len(pattern)-1 {
				return true
			}

			// Try matching ** against zero or more path segments
			for tryPa := pa; tryPa <= len(path); tryPa++ {
				if matchParts(pattern[pi+1:], path[tryPa:]) {
					return true
				}
			}

			return false

		case seg == "*":
			// Matches exactly one segment — just advance both
			pi++
			pa++

		case strings.Contains(seg, "*"):
			// Segment contains inline wildcard (e.g., "openapi*" matches "openapi.json")
			if !matchSegment(seg, path[pa]) {
				return false
			}
			pi++
			pa++

		default:
			if seg != path[pa] {
				return false
			}
			pi++
			pa++
		}
	}

	// Handle trailing ** which can match zero segments
	for pi < len(pattern) && pattern[pi] == "**" {
		pi++
	}

	return pi == len(pattern) && pa == len(path)
}

// matchSegment matches a single path segment against a pattern segment
// containing inline wildcards (e.g., "openapi*" matches "openapi.json").
func matchSegment(pattern, segment string) bool {
	// Simple prefix/suffix matching for inline *
	if strings.HasSuffix(pattern, "*") {
		prefix := strings.TrimSuffix(pattern, "*")
		return strings.HasPrefix(segment, prefix)
	}

	if strings.HasPrefix(pattern, "*") {
		suffix := strings.TrimPrefix(pattern, "*")
		return strings.HasSuffix(segment, suffix)
	}

	// * in the middle: split and check prefix + suffix
	if idx := strings.Index(pattern, "*"); idx >= 0 {
		prefix := pattern[:idx]
		suffix := pattern[idx+1:]
		return strings.HasPrefix(segment, prefix) && strings.HasSuffix(segment, suffix) && len(segment) >= len(prefix)+len(suffix)
	}

	return pattern == segment
}

// ShouldIncludePath evaluates path rules in order and returns whether the
// path should be included. First matching rule wins. If no rule matches,
// the path is included by default.
func ShouldIncludePath(path string, rules []PathRule) bool {
	for _, rule := range rules {
		if MatchPath(rule.Pattern, path) {
			return rule.Action == PathRuleInclude
		}
	}
	return true // default: include
}

// InternalPathRules returns the default rules for excluding framework
// introspection endpoints. Prepend these to user rules when
// ExcludeInternalPaths is enabled.
func InternalPathRules() []PathRule {
	return []PathRule{
		{Pattern: "/", Action: PathRuleExclude},
		{Pattern: "/_farp/**", Action: PathRuleExclude},
		{Pattern: "/_/**", Action: PathRuleExclude},
		{Pattern: "/docs/**", Action: PathRuleExclude},
		{Pattern: "/docs", Action: PathRuleExclude},
		{Pattern: "/openapi*", Action: PathRuleExclude},
		{Pattern: "/asyncapi*", Action: PathRuleExclude},
	}
}

// BuildPathRules combines internal rules (if enabled) with user-defined rules.
// Internal rules are prepended so they run first.
func BuildPathRules(excludeInternal bool, userRules []PathRule) []PathRule {
	if !excludeInternal {
		return userRules
	}
	return append(InternalPathRules(), userRules...)
}

File: farp-ts/src/path-match.ts
import { PathRuleAction } from './types.js';
import type { PathRule } from './types.js';

/**
 * Splits a URL path into segments, stripping leading/trailing slashes.
 */
function splitPath(p: string): string[] {
  const trimmed = p.replace(/^\/+|\/+$/g, '');
  if (trimmed === '') {
    return [];
  }
  return trimmed.split('/');
}

/**
 * Matches a single path segment against a pattern segment containing
 * inline wildcards (e.g., "openapi*" matches "openapi.json").
 */
function matchSegment(pattern: string, segment: string): boolean {
  // Simple prefix matching: "openapi*"
  if (pattern.endsWith('*') && !pattern.startsWith('*')) {
    const prefix = pattern.slice(0, -1);
    return segment.startsWith(prefix);
  }

  // Simple suffix matching: "*json"
  if (pattern.startsWith('*') && !pattern.endsWith('*')) {
    const suffix = pattern.slice(1);
    return segment.endsWith(suffix);
  }

  // * in the middle: "open*json"
  const idx = pattern.indexOf('*');
  if (idx >= 0) {
    const prefix = pattern.slice(0, idx);
    const suffix = pattern.slice(idx + 1);
    return (
      segment.startsWith(prefix) &&
      segment.endsWith(suffix) &&
      segment.length >= prefix.length + suffix.length
    );
  }

  return pattern === segment;
}

/**
 * Recursively matches pattern parts against path parts.
 */
function matchParts(pattern: string[], path: string[]): boolean {
  let pi = 0;
  let pa = 0;

  while (pi < pattern.length && pa < path.length) {
    const seg = pattern[pi];

    if (seg === '**') {
      // If ** is the last pattern segment, it matches everything remaining
      if (pi === pattern.length - 1) {
        return true;
      }

      // Try matching ** against zero or more path segments
      for (let tryPa = pa; tryPa <= path.length; tryPa++) {
        if (matchParts(pattern.slice(pi + 1), path.slice(tryPa))) {
          return true;
        }
      }

      return false;
    } else if (seg === '*') {
      // Matches exactly one segment
      pi++;
      pa++;
    } else if (seg.includes('*')) {
      // Segment contains inline wildcard
      if (!matchSegment(seg, path[pa])) {
        return false;
      }
      pi++;
      pa++;
    } else {
      if (seg !== path[pa]) {
        return false;
      }
      pi++;
      pa++;
    }
  }

  // Handle trailing ** which can match zero segments
  while (pi < pattern.length && pattern[pi] === '**') {
    pi++;
  }

  return pi === pattern.length && pa === path.length;
}

/**
 * Checks whether a URL path matches a glob pattern.
 *
 * Pattern syntax:
 *   - Literal segments match exactly: "/api/v1/users" matches "/api/v1/users"
 *   - "*" matches exactly one path segment: "/api/* /status" matches "/api/users/status"
 *   - "**" matches zero or more path segments: "/api/**" matches "/api", "/api/v1", "/api/v1/users/123"
 */
export function matchPath(pattern: string, path: string): boolean {
  const patternParts = splitPath(pattern);
  const pathParts = splitPath(path);
  return matchParts(patternParts, pathParts);
}

/**
 * Evaluates path rules in order and returns whether the path should be included.
 * First matching rule wins. If no rule matches, the path is included by default.
 */
export function shouldIncludePath(path: string, rules: PathRule[]): boolean {
  for (const rule of rules) {
    if (matchPath(rule.pattern, path)) {
      return rule.action === PathRuleAction.Include;
    }
  }
  return true; // default: include
}

/**
 * Default rules for excluding framework introspection endpoints.
 */
export const INTERNAL_PATH_RULES: PathRule[] = [
  { pattern: '/', action: PathRuleAction.Exclude },
  { pattern: '/_farp/**', action: PathRuleAction.Exclude },
  { pattern: '/_/**', action: PathRuleAction.Exclude },
  { pattern: '/docs/**', action: PathRuleAction.Exclude },
  { pattern: '/docs', action: PathRuleAction.Exclude },
  { pattern: '/openapi*', action: PathRuleAction.Exclude },
  { pattern: '/asyncapi*', action: PathRuleAction.Exclude },
];

/**
 * Combines internal rules (if enabled) with user-defined rules.
 * Internal rules are prepended so they run first.
 */
export function buildPathRules(excludeInternal: boolean, userRules: PathRule[]): PathRule[] {
  if (!excludeInternal) {
    return userRules;
  }
  return [...INTERNAL_PATH_RULES, ...userRules];
}
